File: backend/app/services/puzzle_service.py
import uuid
from datetime import datetime
from typing import List, Optional
from app.schemas.puzzle import (
    PuzzleProblem,
    PuzzleConstraint,
    PuzzleExample,
    PuzzleSubmissionRequest,
    PuzzleSubmissionResponse,
    PuzzleProgress
)

# Demo problem for Milestone 17 Foundation
DEMO_PROBLEM = PuzzleProblem(
    problemId="demo-two-sum",
    title="Two Sum",
    slug="two-sum",
    difficulty="easy",
    category="Arrays",
    description="Given an array of integers `nums` and an integer `target`, return indices of the two numbers such that they add up to `target`.\n\nYou may assume that each input would have exactly one solution, and you may not use the same element twice.\n\nYou can return the answer in any order.",
    constraints=[
        PuzzleConstraint(value="2 <= nums.length <= 10^4"),
        PuzzleConstraint(value="-10^9 <= nums[i] <= 10^9"),
        PuzzleConstraint(value="-10^9 <= target <= 10^9"),
        PuzzleConstraint(value="Only one valid answer exists.")
    ],
    inputFormat="nums = [2,7,11,15]\ntarget = 9",
    outputFormat="[0,1]",
    examples=[
        PuzzleExample(
            input="nums = [2,7,11,15], target = 9",
            output="[0,1]",
            explanation="Because nums[0] + nums[1] == 9, we return [0, 1]."
        ),
        PuzzleExample(
            input="nums = [3,2,4], target = 6",
            output="[1,2]",
            explanation="nums[1] + nums[2] == 6, we return [1, 2]."
        )
    ],
    supportedLanguages=["python", "java", "cpp", "c"],
    status="published",
    sourceType="foundation_demo"
)

class PuzzleService:
    def __init__(self):
        self._problems: List[PuzzleProblem] = [DEMO_PROBLEM]

    def get_all_problems(self) -> List[PuzzleProblem]:
        return self._problems

    def get_problem(self, problem_id: str) -> Optional[PuzzleProblem]:
        for p in self._problems:
            if p.problemId == problem_id:
                return p
        return None
        
    def add_problem(self, problem: PuzzleProblem):
        self._problems.append(problem)

    def submit_solution(self, request: PuzzleSubmissionRequest) -> PuzzleSubmissionResponse:
        # Fake submission for Milestone 17. 
        # Code execution and evaluation will be in Milestone 19.
        submission_id = f"sub_{uuid.uuid4().hex[:10]}"
        return PuzzleSubmissionResponse(
            submissionId=submission_id,
            uid=request.uid,
            problemId=request.problemId,
            language=request.language,
            code=request.code,
            status="pending", # Strictly keeping it pending since we aren't executing it.
            createdAt=datetime.utcnow().isoformat() + "Z"
        )

    def get_progress(self, uid: str) -> PuzzleProgress:
        # Return a foundational empty progress object.
        # This will be populated from Firestore in future milestones.
        return PuzzleProgress(
            uid=uid,
            problemsAttempted=0,
            problemsSolved=0,
            accuracy=0,
            easySolved=0,
            mediumSolved=0,
            hardSolved=0
        )

puzzle_service = PuzzleService()

File: frontend/src/services/puzzleService.ts
import { ApiService } from './api';

export interface PuzzleConstraint {
  value: string;
}

export interface PuzzleExample {
  input: string;
  output: string;
  explanation?: string;
}

export interface PuzzleProblem {
  problemId: string;
  title: string;
  slug: string;
  difficulty: 'easy' | 'medium' | 'hard';
  category: string;
  description: string;
  constraints: PuzzleConstraint[];
  inputFormat: string;
  outputFormat: string;
  examples: PuzzleExample[];
  supportedLanguages: string[];
  status: string;
  sourceType: string;
  tags?: string[];
  expectedComplexity?: Record<string, string>;
  functionSignature?: Record<string, string>;
}

export interface ProblemGenerationRequest {
  category: string;
  difficulty: string;
  problemType: string;
  language: string;
  topic?: string;
}

export interface PuzzleSubmissionRequest {
  uid: string;
  problemId: string;
  language: string;
  code: string;
}

export interface PuzzleSubmissionResponse {
  submissionId: string;
  uid: string;
  problemId: string;
  language: string;
  code: string;
  status: string;
  createdAt: string;
}

export interface CodingSubmissionRequest {
  problemId: string;
  language: string;
  code: string;
}

export interface CodingSubmissionResponse {
  submissionId: string;
  problemId: string;
  status: 'pending' | 'running' | 'accepted' | 'wrong_answer' | 'compile_error' | 'runtime_error' | 'time_limit' | 'memory_limit' | 'system_error';
  language: string;
  passedTests: number;
  totalTests: number;
  score: number;
  executionTimeMs: number;
  memoryUsage: string;
  submittedAt: string;
  isFirstSolve: boolean;
  errorMessage?: string;
}

export interface PuzzleProgress {
  uid: string;
  problemsAttempted: number;
  problemsSolved: number;
  accuracy: number;
  easySolved: number;
  mediumSolved: number;
  hardSolved: number;
}

export const getPuzzleProblems = async (): Promise<PuzzleProblem[]> => {
  return await ApiService.get<PuzzleProblem[]>('/puzzles/problems');
};

export const getPuzzleProblem = async (problemId: string): Promise<PuzzleProblem> => {
  return await ApiService.get<PuzzleProblem>(`/puzzles/problems/${problemId}`);
};

export const submitPuzzleSolution = async (request: PuzzleSubmissionRequest): Promise<PuzzleSubmissionResponse> => {
  return await ApiService.post<PuzzleSubmissionResponse>('/puzzles/submissions', request);
};

export const getPuzzleProgress = async (uid: string): Promise<PuzzleProgress> => {
  return await ApiService.get<PuzzleProgress>(`/puzzles/progress/${uid}`);
};

export const generatePuzzleProblem = async (request: ProblemGenerationRequest): Promise<PuzzleProblem> => {
  return await ApiService.post<PuzzleProblem>('/puzzles/generate', request);
};

export const submitCodingSolution = async (request: CodingSubmissionRequest): Promise<CodingSubmissionResponse> => {
  return await ApiService.post<CodingSubmissionResponse>('/puzzles/submissions', request);
};

export const getSubmissionHistory = async (): Promise<CodingSubmissionResponse[]> => {
  return await ApiService.get<CodingSubmissionResponse[]>('/puzzles/submissions/history');
};
